internal/pkg/pluginengine/change_helper.go

```diff
@@ -166,3 +169,30 @@ func changesForForceDelete(smgr statemanager.Manager, cfg *configloader.Config)
 	}
 	return changes
 }
+
+func renderRefByDependency(tool *configloader.Tool, tools []configloader.Tool, smgr statemanager.Manager) error {
```

Comment:
Too many if in this function, it's hard to read.

Reply:
logic is a bit complicated, it has been explained